cli-impl/src/main/java/org/aya/cli/literate/FaithfulPrettier.java

```diff
@@ -59,19 +49,25 @@ private static void checkHighlights(@NotNull SeqLike<HighlightInfo> highlights)
     var docs = MutableList.<Doc>create();
 
     for (var current : highlights) {
-      var parts = twoKnifeThreeParts(raw, base, current.sourcePos());
-      if (!parts.component1().isEmpty()) docs.append(Doc.plain(parts.component1().toString()));
-      var highlightPart = highlightOne(parts.component2().toString(), current.type());
-      var remainPart = parts.component3();
-      var newBase = parts.component4();
-
-      if (highlightPart != Doc.empty()) {
-        // Hit if: SourcePos contains nothing
-        docs.append(highlightPart);
+      // Cut the `raw` text at `base` offset into three parts: before, current, and remaining,
+      // which needs two split positions: `current.sourcePos().start` and `current.sourcePos().end`, respectively.
+      var knifeCut = twoKnifeThreeParts(raw, base, current.sourcePos());
+      // move forward
+      raw = knifeCut.remaining;
+      base = knifeCut.base;
+
+      // If there's orphan text before the highlighted cut, add it to result as plain text.
+      if (!knifeCut.before.isEmpty()) {
+        // TODO: handle whitespaces in the lexer, and use a new highlight type for them.
+        //  this workaround solution does not work for whitespace in LaTeX.
+        var lines = knifeCut.before.toString().split("\n", -1);
```

Comment:
ij-parsing-core seems not providing detailed whitespace kind, e.g. tabs, spaces, line breaks, etc.

Reply:
Whitespaces can be ~ in LaTeX, and we are not using tabs for now

Reply:
I think it's related to how our lexer is written. We are probably not distinguishing among various whitespace tokens